src/renderer/base/BlinkingState.cpp

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "precomp.h"
+
+#include "../inc/BlinkingState.hpp"
+
+using namespace Microsoft::Console::Render;
+
+// Method Description:
+// - Updates the flag indicating whether cells with the blinking attribute
+//   can animate on and off.
+// Arguments:
+// - blinkingAllowed: true if blinking is permitted.
+// Return Value:
+// - <none>
+void BlinkingState::SetBlinkingAllowed(const bool blinkingAllowed) noexcept
+{
+    _blinkingAllowed = blinkingAllowed;
+    if (!_blinkingAllowed)
+    {
+        _blinkingShouldBeFaint = false;
+    }
+}
+
+// Method Description:
+// - Makes a record of whether the given attribute has blinking enabled or
+//   not, so we can determine whether the screen will need to be refreshed
+//   when the blinking rendition state changes.
+// Arguments:
+// - attr: a text attribute that is expected to be rendered.
+// Return Value:
+// - <none>
+void BlinkingState::RecordBlinkingUsage(const TextAttribute& attr) noexcept
+{
+    _blinkingIsInUse = _blinkingIsInUse || attr.IsBlinking();
+}
+
+// Method Description:
+// - Determines whether cells with the blinking attribute should be rendered
+//   as normal or faint, based on the current position in the blinking cycle.
+// Arguments:
+// - <none>
+// Return Value:
+// - True if blinking cells should be rendered as faint.
+bool BlinkingState::IsBlinkingFaint() const noexcept
+{
+    return _blinkingShouldBeFaint;
+}
+
+// Method Description:
+// - Increments the position in the blinking cycle, toggling the blinking
+//   rendition state on every second call, potentially triggering a redraw of
+//   the given render target if there are blinking cells currently in view.
+// Arguments:
+// - renderTarget: the render target that will be redrawn.
+// Return Value:
+// - <none>
+void BlinkingState::ToggleBlinkingRendition(IRenderTarget& renderTarget) noexcept
+try
+{
+    if (_blinkingAllowed)
+    {
+        // This method is called with the frequence of the cursor blink rate,
+        // but we only want our cells to blink at half that frequency. We thus
+        // have a blinking cycle that loops through four phases...
+        _blinkingCycle = (_blinkingCycle + 1) % 4;
+        // ... and two of those four render the blinking attributes as faint.
+        _blinkingShouldBeFaint = _blinkingCycle >= 2;
+        // Every two cycles (when the state changes), we need to trigger a

```

Comment:
Given that this function is called only after a timer fires, having extra logic here based on the number of times the timer has fired seems slightly awkward since it feels like that logic "should" be incorporated into the timer itself - for terminal this would be easy to do since we already create a separate timer there and so we could just double the tick interval we initialize it with, but on the conhost side this would mean we would create a new timer for SGR 5. If we allow the "cursor blinking is disabled but text blinking is still allowed" use case we would need to create a new timer in conhost anyway, so is this something we should look into?

Reply:
Yeah, it's not ideal, but this little bit of code seemed a lot easier than a whole new timer in conhost, which is quite a chunk of work. But I don't mind be overridden on this if that's the preferred approach. I guess it'll be a matter of cloning most of the `CursorBlinker` class, and then hooking it into the `ConsoleWindowProc` in a couple of places. 

As for disabling cursor blinking while allowing attribute blinking, that's technically still possible, because the timer has to keep running anyway - it's also used for selection scrolling. Although if someone really wanted to disable just the cursor, I'd rather recommend using the DECSET 12 escape sequence. That way we can still use the cursor blink rate to control the actual blinking frequency.

Reply:
Ah, in that case I'm okay with this! Will resolve and someone else can unresolve if they want to chime in

Reply:
Timers are a big deal for `conhost.exe` and one of the places where we get bad marks in the performance gates for introducing more of them or changing their cadence. So I'm actually a big fan of this creativity in re-using the existing timer. 

Normally I'd be for separate timers to separate concerns. But this is appropriate for the situation here, in my opinion.